File: packages/client-ios/Tests/PoppyTests/TestHost.swift
// A reusable PoppyHost stub that records every callback so behavioral tests
// can assert what actions were dispatched, what URLs were rejected, and
// which errors flowed through onError.

import Foundation
@testable import Poppy

final class TestHost: PoppyHost {
    private(set) var actions: [Action] = []
    private(set) var errors: [Error] = []
    var allowURL: (String, ImageContext) -> Bool = { _, _ in true }

    func onAction(_ action: Action) {
        actions.append(action)
    }

    func onError(_ error: Error) {
        errors.append(error)
    }

    func isUrlAllowed(_ url: String, context: ImageContext) -> Bool {
        return allowURL(url, context)
    }
}

File: packages/client-android/src/test/kotlin/dev/poppy/android/TestHost.kt
/*
 * Test PoppyHost that captures dispatched actions for assertion.
 *
 * Used by both BehaviorTest and UpdateSnapshots; the snapshot generator
 * doesn't care about actions but it needs a valid PoppyHost to render. It
 * also overrides isUrlAllowed to accept any URL so the corpus's
 * https://example.com URLs render in tests without needing the network layer
 * (Coil's loader returns a placeholder when offline; that's fine for
 * semantics-tree assertions).
 */

package dev.poppy.android

class TestHost : PoppyHost {
    val actions = mutableListOf<Action>()
    val errors = mutableListOf<Throwable>()

    override fun onAction(action: Action) {
        actions.add(action)
    }

    override fun onError(throwable: Throwable) {
        errors.add(throwable)
    }

    override fun isUrlAllowed(url: String, context: ImageContext): Boolean = true
}
